File: src/main/java/org/apache/hadoop/hbase/master/GroupAdmin.java
/**
 * Copyright 2011 The Apache Software Foundation
 *
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package org.apache.hadoop.hbase.master;

import org.apache.hadoop.hbase.HRegionInfo;

import java.io.IOException;
import java.util.Collection;
import java.util.List;
import java.util.Map;
import java.util.Set;

public interface GroupAdmin {
  /**
   * Get online regions of a region server group.
   *
   * @param groupName the name of the group
   * @return list of online regions this group contains
   */
  List<HRegionInfo> listOnlineRegionsOfGroup(String groupName) throws IOException;

  /**
   * Get member tables of a group.
   *
   * @param groupName the name of the group
   * @return list of table names
   */
  Collection<String> listTablesOfGroup(String groupName) throws IOException;

  /**
   * Gets the group information.
   *
   * @param groupName the group name
   * @return An instance of GroupInfo
   */
  GroupInfo getGroupInfo(String groupName) throws IOException;

  /**
   * Gets the group info of table.
   *
   * @param tableName the table name
   * @return An instance of GroupInfo.
   */
  GroupInfo getGroupInfoOfTable(byte[] tableName) throws IOException;

  /**
   * Move a set of serves to another group
   *
   * @param server the server
   * @param targetGroup the target group
   * @throws java.io.IOException Signals that an I/O exception has occurred.
   * @throws InterruptedException the interrupted exception
   */
  void moveServers(Set<String> server, String targetGroup)
      throws IOException, InterruptedException;


  /**
   * Add a new group
   * @param name name of the group
   * @throws IOException
   */
  void addGroup(String name) throws IOException;

  /**
   * Remove a new group
   * @param name name of the group
   * @throws IOException
   */
  void removeGroup(String name) throws IOException;

  /**
   * Gets the existing groups.
   *
   * @return Collection of GroupInfo.
   */
  List<GroupInfo> listGroups() throws IOException;

  /**
   * Retrieve the GroupInfo a server is affiliated to
   * @param hostPort
   * @return
   * @throws IOException
   */
  GroupInfo getGroupOfServer(String hostPort) throws IOException;

  /**
   * List servers that are currently being moved to a new group
   * @return
   * @throws IOException
   */
  Map<String, String> listServersInTransition() throws IOException;
}

File: src/main/ruby/hbase/group_admin.rb
#
# Copyright 2010 The Apache Software Foundation
#
# Licensed to the Apache Software Foundation (ASF) under one
# or more contributor license agreements.  See the NOTICE file
# distributed with this work for additional information
# regarding copyright ownership.  The ASF licenses this file
# to you under the Apache License, Version 2.0 (the
# "License"); you may not use this file except in compliance
# with the License.  You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
#

include Java
java_import org.apache.hadoop.hbase.util.Pair

# Wrapper for org.apache.hadoop.hbase.client.HBaseAdmin

module Hbase
  class GroupAdmin
    include HBaseConstants

    def initialize(configuration, formatter)
      @admin = org.apache.hadoop.hbase.master.GroupAdminClient.new(configuration)
      @conf = configuration
      @formatter = formatter
    end

    #----------------------------------------------------------------------------------------------
    # Returns a list of groups in hbase
    def listGroups
      @admin.listGroups.map { |g| g.getName }
    end
    #----------------------------------------------------------------------------------------------
    # get a group's information
    def getGroup(group_name)
      group = @admin.getGroup(group_name)
      res = {}
      group.getServers.each do |v|
        if block_given?
          yield(v)
        else
          res += v
        end
      end
    end
    #----------------------------------------------------------------------------------------------
    # add a group
    def addGroup(group_name)
      @admin.addGroup(group_name)
    end
    #----------------------------------------------------------------------------------------------
    # remove a group
    def removeGroup(group_name)
      @admin.removeGroup(group_name)
    end
    #----------------------------------------------------------------------------------------------
    # move server to a group
    def moveServer(dest, *args)
      servers = java.util.TreeSet.new();
      args[0].each do |s|
        servers.add(s)
      end
      @admin.moveServers(servers, dest)
    end
    #----------------------------------------------------------------------------------------------
    # get group of server
    def getGroupOfServer(server)
      @admin.getGroupOfServer(server)
    end
    #----------------------------------------------------------------------------------------------
    # get list tables of groups
    def listTablesOfGroup(group_name)
      @admin.listTablesOfGroup(group_name).map { |g| g }
    end
    #----------------------------------------------------------------------------------------------
    # list servers in transition
    def listServersInTransition()
      iter = @admin.listServersInTransition.entrySet.iterator
      while iter.hasNext
        entry = iter.next
        if block_given?
          yield(entry.getKey, entry.getValue)
        else
          res[entry.getKey] ||= {}
          res[entry.getKey] = entry.getValue
        end
      end
    end
  end
end
